libs/vkd3d/command.c

```diff
@@ -1839,6 +1839,120 @@ static void d3d12_command_list_invalidate_root_parameters(struct d3d12_command_l
+
+        switch (state)
+        {
+            case D3D12_RESOURCE_STATE_VERTEX_AND_CONSTANT_BUFFER:
+                *stages |= queue_shader_stages;
+                *access |= VK_ACCESS_UNIFORM_READ_BIT;
+
+                if (device->bindless_state.flags & VKD3D_BINDLESS_CBV_AS_SSBO)
+                    *access |= VK_ACCESS_SHADER_READ_BIT;
+
+                if (vk_queue_flags & VK_QUEUE_GRAPHICS_BIT)
+                {
+                    *stages |= VK_PIPELINE_STAGE_VERTEX_INPUT_BIT;
+                    *access |= VK_ACCESS_VERTEX_ATTRIBUTE_READ_BIT;
+                }
+                break;
+
+            case D3D12_RESOURCE_STATE_INDEX_BUFFER:
+                *stages |= VK_PIPELINE_STAGE_VERTEX_INPUT_BIT;
+                *access |= VK_ACCESS_INDEX_READ_BIT;
+                break;
+
+            case D3D12_RESOURCE_STATE_RENDER_TARGET:
+                *stages |= VK_PIPELINE_STAGE_COLOR_ATTACHMENT_OUTPUT_BIT;
+                break;
+
+            case D3D12_RESOURCE_STATE_UNORDERED_ACCESS:
+                *stages |= queue_shader_stages;
+                *access |= VK_ACCESS_SHADER_READ_BIT | VK_ACCESS_SHADER_WRITE_BIT;
+                break;
+
+            case D3D12_RESOURCE_STATE_DEPTH_WRITE:
+            case D3D12_RESOURCE_STATE_DEPTH_READ:
+                *stages |= VK_PIPELINE_STAGE_EARLY_FRAGMENT_TESTS_BIT | VK_PIPELINE_STAGE_LATE_FRAGMENT_TESTS_BIT;
+                break;
+
+            case D3D12_RESOURCE_STATE_NON_PIXEL_SHADER_RESOURCE:
+                *stages |= queue_shader_stages & ~VK_PIPELINE_STAGE_FRAGMENT_SHADER_BIT;
+                *access |= VK_ACCESS_SHADER_READ_BIT;
+                break;
+
+            case D3D12_RESOURCE_STATE_PIXEL_SHADER_RESOURCE:
+                *stages |= VK_PIPELINE_STAGE_FRAGMENT_SHADER_BIT;
+                *access |= VK_ACCESS_SHADER_READ_BIT;
+                break;
+
+            case D3D12_RESOURCE_STATE_STREAM_OUT:
+                *stages |= VK_PIPELINE_STAGE_DRAW_INDIRECT_BIT | VK_PIPELINE_STAGE_TRANSFORM_FEEDBACK_BIT_EXT;
+                *access |= VK_ACCESS_TRANSFORM_FEEDBACK_COUNTER_READ_BIT_EXT | VK_ACCESS_TRANSFORM_FEEDBACK_COUNTER_WRITE_BIT_EXT |
+                        VK_ACCESS_TRANSFORM_FEEDBACK_WRITE_BIT_EXT;
+                break;
+
+            case D3D12_RESOURCE_STATE_INDIRECT_ARGUMENT:
+                *stages |= VK_PIPELINE_STAGE_DRAW_INDIRECT_BIT;
+                *access |= VK_ACCESS_INDIRECT_COMMAND_READ_BIT;
+
+                if (device->vk_info.EXT_conditional_rendering)
+                {
+                    *stages |= VK_PIPELINE_STAGE_CONDITIONAL_RENDERING_BIT_EXT;
+                    *access |= VK_ACCESS_CONDITIONAL_RENDERING_READ_BIT_EXT;
+                }
+                break;
+
+            case D3D12_RESOURCE_STATE_COPY_DEST:
+            case D3D12_RESOURCE_STATE_RESOLVE_DEST:
+                *stages |= VK_PIPELINE_STAGE_TRANSFER_BIT;
+                *access |= VK_ACCESS_TRANSFER_WRITE_BIT;
```

Comment:
Transfer ops are also implied barriers.

Reply:
Actually, not for buffers I suppose ...

Reply:
If the resource barrier is on an image, we can probably elide the access mask.

Reply:
We don't take a resource parameter here. Should I add it back?

Reply:
I think so. The implementation will differ depending on resource type.

Reply:
Done.